zcash_client_backend/src/scanning.rs

```diff
@@ -345,42 +327,91 @@ pub(crate) fn scan_block_with_runner<
     prior_block_metadata: Option<&BlockMetadata>,
     mut batch_runner: Option<&mut TaggedBatchRunner<P, K::Scope, T>>,
 ) -> Result<ScannedBlock<K::Nf>, ScanError> {
-    if let Some(scan_error) = check_continuity(&block, prior_block_metadata) {
+    if let Some(scan_error) = check_hash_continuity(&block, prior_block_metadata) {
         return Err(scan_error);
     }
 
     let cur_height = block.height();
     let cur_hash = block.hash();
 
-    // It's possible to make progress without a Sapling tree position if we don't have any Sapling
-    // notes in the block, since we only use the position for constructing nullifiers for our own
-    // received notes. Thus, we allow it to be optional here, and only produce an error if we try
-    // to use it. `block.sapling_commitment_tree_size` is expected to be correct as of the end of
-    // the block, and we can't have a note of ours in a block with no outputs so treating the zero
-    // default value from the protobuf as `None` is always correct.
-    let mut sapling_commitment_tree_size = block
-        .chain_metadata
-        .as_ref()
-        .and_then(|m| {
-            if m.sapling_commitment_tree_size == 0 {
-                None
-            } else {
-                let block_note_count: u32 = block
-                    .vtx
-                    .iter()
-                    .map(|tx| {
-                        u32::try_from(tx.outputs.len()).expect("output count cannot exceed a u32")
-                    })
-                    .sum();
-                Some(m.sapling_commitment_tree_size - block_note_count)
-            }
+    let initial_sapling_tree_size = prior_block_metadata.and_then(|m| m.sapling_tree_size());
+    // If we're below Sapling activation, or Sapling activation is not set, the
+    // tree size is zero
```

Comment:
This comment doesn't seem to match the code. The code sets `nu_sapling_tree_size` to `Some(0)` only when Sapling activation is set and `cur_height` is before it.

Reply:
Ah, right. I think that the code is correct; if Sapling activation is not set, then essentially any operation that attempts to affect the Sapling tree should result in an error, so the comment needs to change.